### ISSUE TITLE
Calling .averages() on a sample that contains a gamma vertex with a tensor crashes

### ISSUE THREAD

**Issue by mhutti1** [improbable-research/keanu#212]
**Describe the bug**
```
 samples.getDoubleTensorSamples(edge.getTotalTimeVertex()).getAverages().lessThan(50).asFlatList().stream().filter(Boolean::booleanValue).count() / (double) TENSOR_SIZE)
```

Causes the following error when I use a Gamma vertex in my project:
```
Exception in thread "main" java.lang.NullPointerException
	at org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.exec(NativeOpExecutioner.java:848)
	at io.improbable.keanu.tensor.INDArrayShim.execBroadcast(INDArrayShim.java:89)
	at io.improbable.keanu.tensor.INDArrayShim.broadcastPlus(INDArrayShim.java:67)
	at io.improbable.keanu.tensor.INDArrayShim.addi(INDArrayShim.java:61)
	at io.improbable.keanu.tensor.dbl.Nd4jDoubleTensor.plusInPlace(Nd4jDoubleTensor.java:558)
	at io.improbable.keanu.tensor.dbl.Nd4jDoubleTensor.plusInPlace(Nd4jDoubleTensor.java:35)
	at java.base/java.util.stream.ReduceOps$1ReducingSink.accept(ReduceOps.java:80)
	at java.base/java.util.Collections$2.tryAdvance(Collections.java:4727)
	at java.base/java.util.Collections$2.forEachRemaining(Collections.java:4735)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.reduce(ReferencePipeline.java:553)
	at io.improbable.keanu.vertices.dbl.DoubleVertexSamples.getAverages(DoubleVertexSamples.java:22)
	at io.improbable.irsim.IRSim.lambda$buildModel$4(IRSim.java:134)
	at io.improbable.irsim.GraphVisualisation.lambda$new$0(GraphVisualisation.java:73)
	at java.base/java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1608)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:658)
	at io.improbable.irsim.GraphVisualisation.<init>(GraphVisualisation.java:61)
	at io.improbable.irsim.IRSim.buildModel(IRSim.java:130)
	at io.improbable.irsim.RoadBinaryParser.complete(RoadBinaryParser.java:150)
	at org.openstreetmap.osmosis.osmbinary.file.BlockInputStream.process(BlockInputStream.java:37)
	at io.improbable.irsim.IRSim.main(IRSim.java:79)
```

**To Reproduce**
I think this could be reproduced by using a Gamma vertex with a tensor and then sampling it and calling ```.getAverages()```

**Expected behavior**
It doesn't null pointer.

**Comment by mhutti1** [comment]
It seems this happens when you try to call .getAverages() on samples with different shapes.

**Comment by mhutti1** [comment]
One issue I had was that creating a ConstantDoubleVertex with a double array creates a 2d 1 by array length tensor not a 1d tensor.

**Comment by mhutti1** [comment]
Yeah so Nd4j.java on line 5080 converts my 1d tensor into a 2d 1 by X tensor and this doesn't match the original shape so they can't be plussed in place.

**Comment by mhutti1** [comment]
Converting my stuff to a 1 by X tensor fixes it.

**Comment by mhutti1** [comment]
This is an example of code that would cause the issue:

'''
HashMap<Long, Vertex> samples = new HashMap<>();
samples.put(1L, Arrays.asList(new ConstantDoubleVertex(new double[] {10, 30}).getValue(), new GammaVertex(new int[] {2}, new ConstantDoubleVertex(1), new ConstantDoubleVertex(2)).getValue()));


new NetworkSamples(samples, 1).getDoubleTensorSamples(1L).getAverages();
'''

It's a bit weird but I guess it depends how we handle tensor dimensionality. I.e do we treat a 2d length 1x2 tensor as equivalent to a 1d length 2 tensor

**Comment by Pinpickle** [comment]
@mhutti1 was correct in that the underlying cause was trying to add two tensors of different shapes. I managed to get a reproduction in Keanu `0.0.12`

```
DoubleTensor.create(new double[] { 10, 30 }, new int[] { 1, 2 }).plusInPlace(DoubleTensor.create(new double[] { 2.1808, 3.4897 }, new int[] { 2 }));
```

However this, and the reproduction case above, do not throw an error in `0.0.13`. From what I can tell this was fixed by #187, which makes adding tensors of different shapes create a new tensor of some sensible shape (even when adding in place).

I'm not sure if the original use-case actually works, but it doesn't throw a null pointer exception anymore. I'll close for now, and if a different issue crops its head trying to do the same thing, please open another issue.
